src/client.ts

```diff
@@ -76,7 +78,128 @@ export default function connect(
+  });
+}
+
+class StandaloneConnection implements Connection {
+  [ALLOW_MODIFICATIONS]: never;
+  private config: NormalizedConnectConfig;
+  private _connection?: ConnectionImpl;
+  private _isClosed: boolean; // For compatibility
+
+  private constructor(config: NormalizedConnectConfig) {
+    this.config = config;
+    this._isClosed = false;
+  }
+  async transaction<T>(
+    action: () => Promise<T>,
+    options?: TransactionOptions
+  ): Promise<T> {
+    return await (await this.connection()).transaction(action, options);
+  }
+  async close(): Promise<void> {
+    try {
+      if (this._connection) {
+        await this._connection.close();
+      }
+      // TODO(tailhook) it makes little sense to close the reconnecting
+      // connection so maybe deprecate this method
+      this._isClosed = true;
+    } finally {
+      this._cleanupProxy();
+    }
+  }
+  private _cleanupProxy(): void {
+    const proxy = proxyMap.get(this);
+    if (proxy != null) {
+      proxy[onConnectionClose]();
+      proxyMap.delete(this);
+    }
+  }
+  isClosed(): boolean {
+    return this._isClosed;
+  }
+  async execute(query: string): Promise<void> {
+    return await (await this.connection()).execute(query);
+  }
+  async query(query: string, args?: QueryArgs): Promise<Set> {
+    return await (await this.connection()).query(query, args);
+  }
+  async queryJSON(query: string, args?: QueryArgs): Promise<string> {
+    return await (await this.connection()).queryJSON(query, args);
+  }
+  async queryOne(query: string, args?: QueryArgs): Promise<any> {
+    return await (await this.connection()).queryOne(query, args);
+  }
+  async queryOneJSON(query: string, args?: QueryArgs): Promise<string> {
+    return await (await this.connection()).queryOneJSON(query, args);
+  }
+  async connection(): Promise<ConnectionImpl> {
+    if (this._isClosed) {
+      throw new errors.InterfaceError("Connection is closed");
+    }
+    if (this._connection && !this._connection.isClosed()) {
+      return this._connection;
+    }
+    const max_time = process.hrtime.bigint()
+      + BigInt(Math.ceil((this.config.waitUntilAvailable || 0) * 1000_000));
+    let iteration = 1;
+    while (true) {
+      for (const addr of this.config.addrs) {
+        try {
+          this._connection = await ConnectionImpl.connect_with_timeout(
+            addr,
+            this.config
+          );
+          return this._connection;
+        } catch (e) {
+          if (e instanceof errors.ClientConnectionError) {
+            if (e.hasTag(errors.SHOULD_RECONNECT)) {
+              if (iteration > 1 && process.hrtime.bigint() > max_time) {
+                throw e;
+              }
+              continue;
+            } else {
+              throw e;
+            }
+          } else {
+            throw e; // this shouldn't happen
+          }
+        }
+      }
+
+      iteration += 1;
+      await sleep(Math.trunc(10 + Math.random() * 200));
```

Comment:
Worst case you'll have 10 milliseconds delay here which feels like way too low for any attempt. I'd suggest changing to 

```suggestion
      await sleep(Math.max(50, Math.random() * 200));
```

But then why isn't `iteration` involved in this function? Ideally attempts should get slightly slower and slower with time.

Reply:
There are two cases where delay is applied:
1. It's "connection refused", which happens via single roundtrip, which on a fast link (and I expect most app -> database links be fast) is the order of `100-1000` micros. Which is orders of magnitude smaller than minimum reconnect of 10 ms, not even saying that it is 100 ms on average. 
2. Connection is dropped by timeout on authentication. That latter is usually in a matter of seconds (we have 60 seconds in current python implementation, should be down to 10 seconds I think, but the order is the same). In this case it doesn't matter how much reconnect delay is as the interval between attempts is dominated by connection timeout anyway.

So my point here is:
1. Minimum interval needed to prevent process from taking 100% CPU and huge network utilization (10 ms is enough)
2. But it should be small enough so that when database is already started we don't wait significantly more before processing requests. I.e. if database starts in 10 seconds, it would be ugly to wait +5 seconds before serving requests, that exponential back-off could add.
3. Randomization is needed just to prevent lots of connections reconnecting simultaneously. And `Math.max(50` is bad for that, because that means 25% of connections will reconnect simultaneously.
4. Our attempts are extremely cheap. This is why we exponential backoff for potentially much expensive things in  `retry` (per RFC1004 spec), but not here.

Reply:
> It's "connection refused", which happens via single roundtrip, which on a fast link (and I expect most app -> database links be fast) is the order of 100-1000 micros. Which is orders of magnitude smaller than minimum reconnect of 10 ms, not even saying that it is 100 ms on average.

Well, 10ms is in the realm of latency between two different clouds or datacenters that's why it caught my attention as a tad too low delay.

I agree with the rest of your comments.

Reply:
Yeah, makes sense to me too.  We can always adjust this later if it turns out to be a poor choice.